### PR TITLE
Add bad character in label check

### PR DESCRIPTION
**Pull request by kumarde**

### CODE REVIEW

File: lints/lint_dnsname_bad_character_in_label.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+	return nil
+}
+
+func (l *DNSNameProperCharacters) CheckApplies(c *x509.Certificate) bool {
+	return true
+}
+
+func labelContainsBadCharacters(domain string) (bool, error) {
+	dnsLabelRegex := "^[A-Za-z0-9*_-]+$"
+	re, err := regexp.Compile(dnsLabelRegex)
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: This should be in init. There's no reason you need to compile this for every check.

File: lints/lint_dnsname_bad_character_in_label.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		return true, err
+	}
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: This is not a lint level failure. There's nothing wrong with the certificate.

File: lints/lint_dnsname_bad_character_in_label.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+		return err
+	}
+	return nil
+}
+
+func (l *DNSNameProperCharacters) CheckApplies(c *x509.Certificate) bool {
+	return true
+}
+
+func labelContainsBadCharacters(domain string, compiledExpression *regexp.Regexp) (bool, error) {
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: Why does this return an error? Error is always `nil`.

File: lints/lint_dnsname_bad_character_in_label.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	"strings"
+)
+
+type DNSNameProperCharacters struct {
+	// Internal data here
+	CompiledExpression *regexp.Regexp
+}
+
+func (l *DNSNameProperCharacters) Initialize() error {
+	dnsLabelRegex := "^[A-Za-z0-9*_-]+$"
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: This should be a constant.

File: lints/lint_dnsname_bad_character_in_label.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		if !compiledExpression.MatchString(label) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (l *DNSNameProperCharacters) RunTest(c *x509.Certificate) (ResultStruct, error) {
+	if c.Subject.CommonName != "" {
+		badCharacterFound, err := labelContainsBadCharacters(c.Subject.CommonName, l.CompiledExpression)
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: Why isn't this just a helper method that has access to `l.CompiledExpression`?

File: lints/lint_dnsname_bad_character_in_label.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	"regexp"
+	"strings"
+)
+
+type DNSNameProperCharacters struct {
+	// Internal data here
+	CompiledExpression *regexp.Regexp
+}
+
+func (l *DNSNameProperCharacters) Initialize() error {
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: The code base never actually checks whether this call returns an error.

Comment (dadrian) [reply lints/lint_dnsname_bad_character_in_label.go]: IMHO `Initialize()` should not return an error, similar to `func init()`. If some really bad happens, `panic()` internally. There's not much the framework could do with a non-nil error, anyway.

Comment (zakird) [reply lints/lint_dnsname_bad_character_in_label.go]: I'm fine with that. We just gotta do something about it.

Comment (kumarde) [reply lints/lint_dnsname_bad_character_in_label.go]: `MustCompile` does exactly this. I'll change to that.

File: lints/lint_dnsname_bad_character_in_label.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	const dnsLabelRegex = "^[A-Za-z0-9*_-]+$"
+	l.CompiledExpression = regexp.MustCompile(dnsLabelRegex)
+	return nil
+}
+
+func (l *DNSNameProperCharacters) CheckApplies(c *x509.Certificate) bool {
```
Comment (zakird) [inline lints/lint_dnsname_bad_character_in_label.go]: This seems like it should only apply to certificates with domains. I don't think this applies to CA names, etc.

**Pull request merged**